metadata/src/main/java/org/apache/kafka/controller/QuorumController.java

```diff
@@ -1455,30 +1523,56 @@ private void replay(ApiMessage message, Optional<OffsetAndEpoch> snapshotId, lon
     }
 
     private void maybeGenerateSnapshot() {
-        if (committedBytesSinceLastSnapshot >= snapshotMaxNewRecordBytes &&
-            snapshotGeneratorManager.generator == null
-        ) {
-            if (!isActiveController()) {
-                // The active controller creates in-memory snapshot every time an uncommitted
-                // batch gets appended. The in-active controller can be more efficient and only
-                // create an in-memory snapshot when needed.
-                snapshotRegistry.getOrCreateSnapshot(lastCommittedOffset);
+        if (!snapshotGeneratorManager.snapshotInProgress()) {
+            Set<SnapshotReason> snapshotReasons = new HashSet<>();
+            // Check if a snapshot should be generated because of committed bytes
+            if (committedBytesSinceLastSnapshot >= snapshotMaxNewRecordBytes) {
+                snapshotReasons.add(
+                    SnapshotReason.maxBytesExceeded(committedBytesSinceLastSnapshot, snapshotMaxNewRecordBytes)
+                );
             }
 
-            log.info(
-                "Generating a snapshot that includes (epoch={}, offset={}) after {} committed bytes since the last snapshot because, {}.",
-                lastCommittedEpoch,
-                lastCommittedOffset,
-                committedBytesSinceLastSnapshot,
-                SnapshotReason.MaxBytesExceeded
-            );
+            // Check if a snapshot should be generated because of committed append times
+            if (maxSnapshotIntervalMs > 0) {
+                // Time base snasphots are enabled
+                long snapshotIntervalMs = time.milliseconds() - oldestCommittedLogOnlyAppendTimestamp;
+                if (snapshotIntervalMs >= maxSnapshotIntervalMs) {
+                    snapshotReasons.add(SnapshotReason.maxIntervalExceeded(snapshotIntervalMs, maxSnapshotIntervalMs));
+                } else {
+                    maybeScheduleNextGenerateSnapshot();
+                }
+            }
 
-            snapshotGeneratorManager.createSnapshotGenerator(
-                lastCommittedOffset,
-                lastCommittedEpoch,
-                lastCommittedTimestamp
-            );
-            committedBytesSinceLastSnapshot = 0;
+            if (!snapshotReasons.isEmpty()) {
+                if (!isActiveController()) {
+                    // The active controller creates in-memory snapshot every time an uncommitted
```

Comment:
Nit:
Would be a little easier to read if the comment was rephrased to put what is happening here:
```
// The standby controllers do not create in-memory snapshots every time a 
// batch gets appended, so we create it now.
```

Reply:
I reorder the sentences and improved the wording. I kept both sentences because I think they are important for anyone interested in understanding when the controllers generates and deletes in-memory snapshots.